### PR TITLE
API: raise when using unaligned Series as a method argument

### LINKED ISSUE #3256: API: Series alignment for non binary op methods
We have a few geo methods which take an additional arraylike input as a parameter; `interpolate`, `offset_curve` and `voronoi_polygons`, but we don't have a consistent way of dealing with alignment between `self.geometry` and the passed arraylike when it is a Series.

>Just noticing we have a bit of inconsistency of how to handle when we have a non binary op method expecting an arraylike. For binary ops, we  warn on alignment, here we auto align, in `interpolate`, distance throws a ValueError if it does not align, and for `offset_curve` we just feed unaligned GeometryArray to shapely.

I would prefer we warn or throw on unaligned data, but whatever we decide we should be consistent across similar situations.

_Originally posted by @m-richards in https://github.com/geopandas/geopandas/pull/3177#discussion_r1570298342_

### PR DESCRIPTION
**Pull request by martinfleis**
Closes #3256

### CODE REVIEW

File: geopandas/base.py [hunk 1]
```diff
@@ -88,13 +88,23 @@ def _delegate_property(op, this):
         return Series(data, index=this.index)
 
 
-def _delegate_geo_method(op, this, *args, **kwargs):
+def _delegate_geo_method(op, this, **kwargs):
```
Comment (martinfleis) [inline geopandas/base.py:91]: Restricting this to `**kwargs` only ensures we always have a key to be printed in the error message.

File: geopandas/base.py [hunk 2]
```diff
@@ -88,13 +88,23 @@ def _delegate_property(op, this):
...
 
+    for key, val in kwargs.items():
+        if isinstance(val, pd.Series) and not val.index.equals(this.index):
+            raise ValueError(
+                f"Index of the Series passed as '{key}' does not match index of the "
+                "GeoSeries. If you want both Series to be aligned, align them before "
+                f"passing them to this method as `this, {key} = this.align({key})`. If "
+                f"you want to ignore the index pass the underlying array as '{key}' "
+                f"using `{key}.values`."
+            )
```
Comment (m-richards) [inline geopandas/base.py:104]: So `this` is not necessarily a GeoSeries (contrary to the type comment above), it's just a subtype of GeopandasBase, so referring to GeoSeries in the body is potentially misleading if `gdf.concave_hull(ratio=...)` is called. I'd suggest we do something like this to try and be a bit more friendly:
```suggestion
    from .geodataframe import GeoDataFrame

    if isinstance(this, GeoSeries):
        klass, var_name = "GeoSeries", "g"
    elif isinstance(this, GeoDataFrame):
        klass, var_name = "GeoDataFrame", "gdf"
    else:
        klass, var_name = this.__class__.__name__, "this"

    for key, val in kwargs.items():
        if isinstance(val, pd.Series) and not val.index.equals(this.index):
            raise ValueError(
                f"Index of the Series passed as '{key}' does not match index of the "
                f"{klass}. If you want both Series to be aligned, align them before "
                f"passing them to this method as "
                f"`{var_name}, {key} = {var_name}.align({key})`. If "
                f"you want to ignore the index, pass the underlying array as '{key}' "
                f"using `{key}.values`."
            )
```
(we could also not refer to GeoSeries/GeoDataFrame at all and just refer to this, but I think there's value in trying to have more descriptive language for the common case.)

**Review by m-richards — APPROVED** [review]
Thanks Martin, left a comment on the actual error message, but otherwise makes sense to me. Just needs a changelog note.

**Review by jorisvandenbossche — APPROVED** [review]
Looks good, thanks!

**Pull request merged**